### ISSUE TITLE
docs(docs-infra): fix (notation of) microsyntax grammar

### ISSUE THREAD

**Issue by maiermic** [angular/angular#33436]
# 📚 Docs or angular.io bug report

### What's the affected URL?
https://angular.io/guide/structural-directives#grammar

### Description

The current grammar is defined by the following production rules:

    *:prefix="( :let | :expression ) (';' | ',')? ( :let | :as | :keyExp )*"
    keyExp = :key ":"? :expression ("as" :local)? ";"?
    let = "let" :local "=" :export ";"?
    as = :export "as" :local ";"?

I describe several issues of the used notation and grammar rules. Each issue and my suggested solution is described in a section. The sections are separated by a horizontal line in the following. The presented solutions build on the (changes of the) previous solutions. The last section shows the final result (production rules).

---

**Issue:** Production rule `*:prefix` is not formatted consistently.
**Solution:** Use one space character before and after `=`. Use one space character before `(` and no space character after `(`. Use no space character before `)`. Use one space character after `)` if it is not followed by an operator (`?` or `*`). Otherwise, only write one space character after the operator, but only if the operator is not followed by `)`.

    *:prefix = " (:let | :expression) (';' | ',')? (:let | :as | :keyExp)* "

---

**Issue:** Prefix `:` of nonterminal symbols creates no added value.
**Solution:** Remove `:`

    *prefix = " (let | expression) (';' | ',')? (let | as | keyExp)* "
    keyExp = key ":"? expression ("as" local)? ";"?
    let = "let" local "=" export ";"?
    as = export "as" local ";"?

---

**Issue:** No consistent use of double quotation marks `"` in the production rules. `"` is used as terminal symbol in production rule `*prefix`, but as delimiter of string literals (sequence of terminal symbols) in the other production rules `keyExp`, `let` and `as`.
**Solution:** Use `'` as delimiter and surround the double quotation terminal symbol with delimiters:

    *prefix = '"' (let | expression) (';' | ',')? (let | as | keyExp)* '"'
    keyExp = key ':'? expression ('as' local)? ';'?
    let = 'let' local '=' export ';'?
    as = export 'as' local ';'?

---

**Issue:** Assignment of `export` to `local` variable is mandatory in production rule `let`, but the implementation allows to omit explicit assignment, like `let x` in

    <p *ngFor="let x of ['a', 'b']">{{x}}</p>

which is treated like an implicit assignment of the special context property `$implicit` to `x`.

**Solution:** Make assignment optional in production rule `let`:

    let = 'let' local ('=' export)? ';'?

---

**Issue:** The current grammar definition allows the production of two consecutive semicolons, see derivation example:

    *prefix = '"' (let | expression) (';' | ',')? (let | as | keyExp)* '"'
            = '"' (let             )  ';'         (let | as | keyExp)* '"'
            = '"' ('let' local ('=' export)? ';'?) ';' (let | as | keyExp)* '"'
            = '"'  'let' local  '=' export   ';'   ';' (let | as | keyExp)* '"'

But the current implementation throws `Template parse errors` in that case, for example:

    <p *ngFor="let i = index;; let x of ['a', 'b']">{{i}}: {{x}}</p>

**Solution:** Extract semicolon production from rule `let` into rule `*prefix` and move separator production `(';' | ',')?` to the alternative `expression`:

    *prefix = '"' (let ';'? | expression (';' | ',')?) (let ';'? | as | keyExp)* '"'
    keyExp = key ':'? expression ('as' local)? ';'?
    let = 'let' local ('=' export)?
    as = export 'as' local ';'?

---

**Issue:** Production rules do not consider that (terminal symbol) colon may be used instead of semicolon separator. For example

    <p *ngFor="let item of [1,2,3] as items; trackBy: myTrack, index as i">{{i}}: {{item}}</p>

can not be produced using the current rules due to the colon after `myTrack`, but it is valid syntax in the implementation.

**Solution:** Define production rule `sep` (abbreviation of *separator*):

    *prefix = '"' (let sep? | expression sep?) (let sep? | as | keyExp)* '"'
    sep = ';' | ','
    keyExp = key ':'? expression ('as' local)? sep?
    let = 'let' local ('=' export)?
    as = export 'as' local sep?

This can be simplified by moving `sep?` around in the definitions

    *prefix = '"' (let | expression) sep? ((let | as | keyExp) sep?)* '"'
    sep = ';' | ','
    keyExp = key ':'? expression ('as' local)?
    let = 'let' local ('=' export)?
    as = export 'as' local

This code block represents my **final suggestion**.

**Comment by maiermic** [comment]
@gkalpak left a [comment](https://github.com/angular/angular/pull/31517#pullrequestreview-260938335)
> OOC, why use the `:`-prefix? It is is a little confusing given that `:` can be part of DSL.

 in the pull request that introduced the grammar documentation, but the author of the pull request @kapunahelewong did not answer that question back then. It must have been overlooked.

**Comment by trotyl** [comment]
Additional issues:

+ HTML attribute value doesn't have to use double quote (`"`), one can always use single quote (`'`), or even no quote if no whitespace involved;
+ Whitespace that being a part of the syntax should be explicitly specified;
+ `:export` before `"as" :local` is also an optional part, one can write `*ngIf="foo as bar as baz as blah"`;
+ `("as" :local)?` in never a part of `keyExp`, `*ngIf="foo as bar"` is treated as `*ngIf="foo; as bar;"` after ASI, just two consecutive but unrelated statements;

**Comment by maiermic** [comment]
> * HTML attribute value doesn't have to use double quote (`"`), one can always use single quote (`'`), or even no quote if no whitespace involved;

@trotyl Which part of the documentation states the syntax of HTML attributes in general? There have to be more constraints than no-whitespace for expressions if no quotes are used. For example, operator `/` leads to `Template parse errors`

```html
<p *ngIf=isNaN(2/'foo')>expression contains no whitespace</p>
```

We could write

```
*prefix = ('"' microExpr '"') | ("'" microExpr "'") | attrValueExpr
microExpr = (let | localExpr) sep? ((let | as | keyExp) sep?)*
```

and add a description of `attrValueExpr` (in the table below the production rule `*prefix`) that describes the constraints in words.


> * Whitespace that being a part of the syntax should be explicitly specified;

I wonder if there is a concise way of writing this and if it creates enough additional value to justify the more verbose rules it causes. I guess it might be good enough to ignore whitespace in the production rules, even though you could argue that it is not correct. The actual syntax should be still clear to most readers without it. Or are there any unusual places where whitespace can occur that you've come across? I understand it the way you can treat whitespace in ANTL4 (see [this](https://stackoverflow.com/a/15503680/1065654) answer on StackOverflow). You should also consider that the production rules of some nonterminal symbols (`prefix`, `key`, `local`, `export` and `expression`) are only described in words. I do not assume that the goal of the given grammar is to be absolutely precise/accurate.

> * `("as" :local)?` may not belong to any `keyExp` in single expression directives like `*ngIf="foo as bar"`;

How about this change (see `localExpr`) based on the current final suggestion?

```
*prefix = '"' (let | localExpr) sep? ((let | as | keyExp) sep?)* '"'
sep = ';' | ','
keyExp = key ':'? localExpr
localExpr = expression ('as' local)?
let = 'let' local ('=' export)?
as = export 'as' local
```

**Comment by trotyl** [comment]
> Which part of the documentation states the syntax of HTML attributes in general?

That's the exact problem, HTML attribute syntax is not mentioned at all but a specific form is used. It may misguide people about how HTML syntax is defined.

> There have to be more constraints than no-whitespace for expressions if no quotes are used. For example, operator / leads to Template parse errors

It has nothing to do with `/`, the problem is you have used quotes (`'foo'`) in the literal but in wrong place.

> `*prefix = ('"' microExpr '"') | ("'" microExpr "'") | attrValueExpr`

That's wrong syntax, no whitespace is allowed around the `=` punctuator (that's not a production definition). What's more, it totally mess up literals and named productions, should be something like:

```
"*" NWS prefix NWS "=" NWS ('"' microExpr '"') | ("'" microExpr "'") | attrValueExpr
```

But which is not expressive at all, so I'd suggest don't cover HTML attribute syntax itself, but only the syntax inside attribute value, namely the Angular microsyntax.

> You should also consider that the production rules of some nonterminal symbols (prefix, key, local, export and expression) are only described in words.

It definitely should be documented there, people can actually write:

```html
<p **foo="'abc'"></p>
```

With the definition:

```typescript
@Directive({
  selector: `[*foo]`
})
export class Foo {
  @Input('*foo') foo: string
}
```

https://stackblitz.com/edit/angular-bxzuvd?file=src%2Fapp%2Fapp.component.html

It's totally valid even looks weird.

> How about this change (see localExpr) based on the current final suggestion?

It looks much better.

**Comment by kapunahelewong** [comment]
> @gkalpak left a [comment](https://github.com/angular/angular/pull/31517#pullrequestreview-260938335)
> 
> > OOC, why use the `:`-prefix? It is is a little confusing given that `:` can be part of DSL.
> 
> in the pull request that introduced the grammar documentation, but the author of the pull request @kapunahelewong did not answer that question back then. The comment/question must have been overlooked.

@maiermic, thanks for the context. 😊If I remember correctly, that conversation ended up offline and we chose to go with that, but I don't remember the details as to why. It may be that it also was ultimately overlooked on my part. Anyway, I'd be happy to fix things, but I'm not as versed in this material as you and @trotyl, and it might take me a while before I can get into it. Would you be interested in submitting a PR or advising me on one when I can do it?

**Comment by maiermic** [comment]
@kapunahelewong Thanks, if nobody remembers why the `:`-prefix has been used and nobody makes a point why it is beneficial to keep it, I will leave it in my final suggestion. I will create a PR when the discussion comes to an end. The commit messages will document why a change has been made. It might make sense to involve potential reviewers in the discussion. @trotyl's insights really helped me to improve my solution.

**Comment by maiermic** [comment]
> > Which part of the documentation states the syntax of HTML attributes in general?
> 
> That's the exact problem, HTML attribute syntax is not mentioned at all but a specific form is used. It may misguide people about how HTML syntax is defined.

I agree, but that should be covered in another issue.

> > There have to be more constraints than no-whitespace for expressions if no quotes are used. For example, operator / leads to Template parse errors
> 
> It has nothing to do with `/`, the problem is you have used quotes (`'foo'`) in the literal but in wrong place.

The quotes of the string `'foo'` are another problem. The operator `/` leads also to `Template parse errors`

```
<p *ngIf=isNaN(2/1)>expression contains no whitespace</p>
```

> > `*prefix = ('"' microExpr '"') | ("'" microExpr "'") | attrValueExpr`
> 
> That's wrong syntax, no whitespace is allowed around the `=` punctuator (that's not a production definition). What's more, it totally mess up literals and named productions, should be something like:
> 
> ```
> "*" NWS prefix NWS "=" NWS ('"' microExpr '"') | ("'" microExpr "'") | attrValueExpr
> ```
> 
> But which is not expressive at all, so I'd suggest don't cover HTML attribute syntax itself, but only the syntax inside attribute value, namely the Angular microsyntax.

You are totally right and I agree with your conclusion.

> > You should also consider that the production rules of some nonterminal symbols (prefix, key, local, export and expression) are only described in words.
> 
> It definitely should be documented there, people can actually write:
> 
> ```
> <p **foo="'abc'"></p>
> ```
> 
> With the definition:
> 
> ```ts
> @Directive({
>   selector: `[*foo]`
> })
> export class Foo {
>   @Input('*foo') foo: string
> }
> ```
> 
> https://stackblitz.com/edit/angular-bxzuvd?file=src%2Fapp%2Fapp.component.html
> 
> It's totally valid today even looks weird. To prevent user from doing so this kind of chars should be explicitly forbid from the production.

I agree, but this affects the implementation. Hence, it should be addressed in another issue.

**Comment by maiermic** [comment]
@trotyl How about this suggestion (based on my latest suggestion at the end of [this comment](https://github.com/angular/angular/issues/33436#issuecomment-547147217)) that

> cover HTML attribute syntax itself, but only the syntax inside attribute value, namely the Angular microsyntax

and considers whitespace?

```
microSyntax = (let | localExpr) (sepWs (let | as | keyExp))*
ws = ' ' | '\t' | '\r' | '\n'
sep = ';' | ','
sepWs = ws* (ws | sep) ws*
keyExp = key ws* (ws | ':') ws* localExpr
localExpr = expression (ws+ 'as' ws+ local)?
let = 'let' ws+ local (ws* '=' ws* export)?
as = export ws+ 'as' ws+ local
```

**Comment by trotyl** [comment]
> The quotes of the string 'foo' are another problem. The operator / leads also to Template parse errors

Thanks for clarifying, it seems to be a BUG of Angular HTML parser, [`/` is indeed allowed in unquote attribute value](https://html.spec.whatwg.org/multipage/parsing.html#attribute-value-(unquoted)-state).

> and considers whitespace?

Personally I believe it should be defined outside microsyntax, as it's can be used in any template expression or statement.

**Comment by maiermic** [comment]
Do you mean the `ws` rule should be defined outside microsyntax? Do you mean outside of the [Microsyntax topic](https://angular.io/guide/structural-directives#microsyntax)? Where should that be?  Do you like to add a grammar to the topic [Template Syntax](https://angular.io/guide/template-syntax)? Do you like to have a complete grammar in the documentation?

**Comment by kapunahelewong** [comment]
Docs team would like engineering input as to the preferred microsyntax in this doc along with info on why it's that way.

**Comment by maiermic** [comment]
What do you mean by "engineering input"? To whom is your request addressed?

**Comment by kapunahelewong** [comment]
Hi @maiermic, this is a label that the documentation team uses that means that a writer will need the help of an engineer to take on the documentation issue. Last week, we were going through issues doing several things. We needed to add the `triage#1` label (to show that we had triaged the issue in this particular effort to get through the outstanding issues), `needs eng input` (if the person doing the triage thinks that the docs team member needs engineering assistance, that is a subject matter expert), and to leave a comment so the other members of the docs team would know the status of that issue. 

If we need engineering input, then the ultimate goal is for a writer to be paired with one of the engineers on the team to make sure that we create accurate documentation on a given point. So, we're just trying to organize and get a handle on the backlog.

Does that help? Thanks for asking!
